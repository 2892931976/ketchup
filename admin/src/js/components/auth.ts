let store = require('store/dist/store.modern') as StoreJSStatic;
import * as m from 'mithril';
import * as Toaster from 'components/toaster';

export interface User {
  email: string;
  uuid: string;
}

interface Preferences {
  hideMenu: boolean;
}

interface Storer {
  get(key: string): any;
  set(key: string, val: any): void;
}

class DummyStore {
  data: any;

  constructor() {
    this.data = {};
  }

  get(key: string): any {
    return this.data[key];
  }

  set(key: string, val: any) {
    this.data[key] = val;
  }
}

let cachedUser: User = null;
let dummyStore = new DummyStore();

// AuthController is a super class for controllers which may require auth
export class AuthController {
  user: User;
  _userPromise: Promise<User>;
  store: Storer;

  constructor(user: User = null) {
    this.store = store.disabled ? dummyStore : store;
    this.user = user || cachedUser;
    if (this.user) {
      this._userPromise = new Promise<User>((resolve) => {
        resolve(this.user);
      });
    } else {
      this._userPromise = m.request({
        method: 'GET',
        url: '/api/v1/user',
        background: false,
      }).then((res: User) => {
        if (!res.uuid) {
          throw new Error('not logged in');
        }
        this.user = cachedUser = res;
        return res;
      }).catch(() => {
        cachedUser = null;
        return null;
      });
    }
  }

  logout() {
    m.request({
      method: 'GET',
      url: '/api/v1/logout',
      background: false
    }).then(() => {
      Toaster.add('logged out');
      setTimeout(() => {
        location.reload();
      }, 2000);
    });
  }

  private get storeKey(): string {
    if (!this.user) {
      return null;
    }
    return `user-${this.user.uuid}`;
  }

  private getPrefs(): Preferences {
    if (!this.user) {
      return null;
    }
    let prefs: Preferences = this.store.get(this.storeKey);
    if (!prefs) {
      prefs = {
        hideMenu: false,
      };
      this.store.set(this.storeKey, prefs);
    }
    return prefs;
  }

  pref<K extends keyof Preferences>(key: K): Preferences[K] {
    if (!this.user) {
      return; // error
    }
    let prefs: Preferences = this.getPrefs();
    if (!prefs) {
      return; // error
    }
    return prefs[key];
  }

  setPref<K extends keyof Preferences>(key: K, val: Preferences[K]) {
    if (!this.user) {
      return; // error
    }
    let prefs: Preferences = this.getPrefs();
    prefs[key] = val;
    this.store.set(this.storeKey, prefs);
  }
}

export class MustAuthController extends AuthController {
  ready: boolean;
  constructor() {
    super();
    this.ready = false;
    this._userPromise.then((user) => {
      if (!user || !user.uuid) {
        window.location.assign(`/admin/login?next=${window.location.pathname}`);
      } else {
        this.ready = true;
      }
    });
  }
}
