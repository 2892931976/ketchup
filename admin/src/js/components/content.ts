import * as m from 'mithril';
import * as API from 'lib/api';
import Page from 'lib/page';
import EditorComponent from 'components/editors/editor';
import QuillComponent from 'components/editors/html';
import TextEditorComponent from 'components/editors/text';
import CodeMirrorComponent from 'components/editors/markdown';

const control = (...el: any[]) => m('.controls', m('.control.control-full', el));

export function renderEditor(page: Page, c: API.Content, hideLabel: boolean) {
  switch (true) {
    case c.short != null:
      return shortEditor(page, c, hideLabel);
    case c.text != null:
      return textEditor(page, c, hideLabel);
    default:
      console.log('warning: no editor defined for object:', c);
      c.text = { type: 'html' }; // set a default
      return textEditor(page, c, hideLabel);
  }
}

// textEditor is an editor for a large block of text
function textEditor(_: Page, c: API.Content, hideLabel: boolean = false) {
  switch (c.text.type) {
    case 'html':
      return control(
        hideLabel ? '' : m('.label', c.key),
        m(EditorComponent, { content: c })
      );

    case 'markdown':
      return control(
        hideLabel ? '' : m('.label', c.key),
        m(CodeMirrorComponent, { content: c }) // todo: assign id?
      );

    case 'text':
      return control(
        hideLabel ? '' : m('.label', c.key),
        m(TextEditorComponent, { content: c }) // todo: assign id?
      );

    default:
      return null;
  }
}

function shortEditor(page: Page, c: API.Content, hideLabel: boolean = false) {
  switch (c.short.type) {
    case 'html':
      return control(
        hideLabel ? '' : m('.label', c.key),
        m(QuillComponent, { content: c })
      );

    case 'text':
    case 'markdown':
      return control(
        hideLabel ? '' : m('.label', c.key),
        m('input[type=text]', {
          oncreate: (v: Mithril.VnodeDOM<any, any>) => {
            (v.dom as HTMLTextAreaElement).value = c.value || '';
          },
          onchange: (e: EventTarget) => {
            c.value = (e as any).target.value;
            page.updateContent(c);
          }
        })
      );

    default:
      return null;
  }
}