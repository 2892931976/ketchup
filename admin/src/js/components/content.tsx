import * as React from 'react';
import * as API from 'lib/api';
import EditorComponent from 'components/editors/editor';
import QuillComponent from 'components/editors/html';
import TextEditorComponent from 'components/editors/text';
import CodeMirrorComponent from 'components/editors/markdown';

interface ContentEditor {
  shouldRender(content: API.Content): boolean;
  render(content: API.Content): React.ReactElement<any>;
}

let LongHTMLEditor: ContentEditor = {
  shouldRender: (content: API.Content) => content.text != null && content.text.type == 'html',
  render: (content: API.Content) => <EditorComponent content={content} />,
};

let LongMarkdownEditor: ContentEditor = {
  shouldRender: (content: API.Content) => content.text != null && content.text.type == 'markdown',
  render: (content: API.Content) => <CodeMirrorComponent content={content} />,
};

let LongTextEditor: ContentEditor = {
  shouldRender: (content: API.Content) => content.text != null && content.text.type == 'text',
  render: (content: API.Content) => <TextEditorComponent content={content} />,
};

// exported for test
export class ShortTextEditorComponent extends React.PureComponent<{content: API.Content}> {
  textInput: React.RefObject<HTMLInputElement>;
  constructor(props: any) {
    super(props);
    this.textInput = React.createRef();
  }
  componentDidMount() {
    this.textInput.current.value = this.props.content.value;
  }

  render() {
    return (
      <input
        type="text"
        ref={this.textInput}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          this.props.content.value = e.target.value;
        }}
      />
    );
  }
}

let ShortTextEditor: ContentEditor = {
  shouldRender: (content: API.Content) =>
    content.short != null && (content.short.type == 'text' || content.short.type == 'markdown'),
  render: (content: API.Content) => <ShortTextEditorComponent content={content} />,
};

let ShortHTMLEditor: ContentEditor = {
  shouldRender: (content: API.Content) => content.short != null && content.short.type == 'html',
  render: (content: API.Content) => <QuillComponent content={content} />,
};

let editors: ContentEditor[] = [
  LongHTMLEditor,
  LongMarkdownEditor,
  LongTextEditor,
  ShortHTMLEditor,
  ShortTextEditor,
];

export let Editor: React.SFC<{
  content: API.Content | API.Data;
  hideLabel: boolean;
}> = (props) => {
  for (var i = 0; i < editors.length; i++) {
    let editor = editors[i];
    if (editor.shouldRender(props.content)) {
      return (
        <div className="controls">
          <div className="control control-full">
            {props.hideLabel ? '' : <div className="label">{props.content.key}</div>}
            {editor.render(props.content)}
          </div>
        </div>
      );
    }
  }
};
