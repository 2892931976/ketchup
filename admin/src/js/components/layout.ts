import Navigation from './navigation';

export default (content: string | Mithril.VirtualElement) =>
  m('.container', [
    m.component(Navigation),
    m('.container--body', content)
  ]);
