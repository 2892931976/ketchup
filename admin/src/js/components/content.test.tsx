import { assert } from 'chai';
import * as mq from 'mithril-query';
import * as API from 'lib/api';
import { renderEditor } from 'components/content';
import CodeMirrorComponent from 'components/editors/markdown';
import TextEditorComponent from 'components/editors/text';
import EditorComponent from 'components/editors/editor';

describe('renderEditor', function() {
  function testEditor(content: API.Content, editorComponent: any) {
    let out = mq(renderEditor(content, false));
    assert.containSubset(out.rootNode, {
      tag: 'div',
      attrs: { className: 'controls' },
      children: [{
        attrs: { className: 'control control-full', },
        children: [{
          attrs: { className: 'label' },
          text: content.key
        }, {
          attrs: { content }
        }]
      }]
    });

    let editor = out.find('.control-full')[0].children[1];
    assert.equal(editor.tag, editorComponent);
    assert.strictEqual(editor.attrs.content, content);

    out = mq(renderEditor(content, true));
    assert.equal(out.find('.label').length, 0);
  }

  it('text:markdown -> LongMarkdownEditor', function() {
    testEditor({
      key: 'akey',
      value: '*hello world*',
      text: {
        title: 'thetitle',
        type: 'markdown',
      }
    }, CodeMirrorComponent);
  });

  it('text:text -> LongTextEditor', function() {
    testEditor({
      key: 'akey',
      value: '*hello world*',
      text: {
        title: 'thetitle',
        type: 'text',
      }
    }, TextEditorComponent);
  });

  it('text:html -> LongHTMLEditor', function() {
    testEditor({
      key: 'akey',
      value: '*hello world*',
      text: {
        title: 'thetitle',
        type: 'html',
      }
    }, EditorComponent);
  });

  it('short:markdown -> ShortMarkdownEditor', function() {
    let content: API.Content = {
      key: 'akey',
      value: '*hello world*',
      short: {
        title: 'thetitle',
        type: 'markdown',
      }
    }
    let out = mq(renderEditor(content, false));
    let handler = out.find('input')[0].attrs.onchange;
    assert.isFunction(handler);
    let event = { target: { value: 'goodbye moon' }};
    handler(event)
    assert.equal(content.value, event.target.value)
  });
});
