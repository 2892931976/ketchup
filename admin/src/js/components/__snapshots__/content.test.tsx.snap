// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor> short:markdown -> ShortMarkdownEditor 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    content={
      Object {
        "key": "akey",
        "short": Object {
          "title": "thetitle",
          "type": "markdown",
        },
        "value": "*hello world*",
      }
    }
    hideLabel={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="control control-full"
      >
        <div
          className="label"
        >
          akey
        </div>
        <ShortTextEditorComponent
          content={
            Object {
              "key": "akey",
              "short": Object {
                "title": "thetitle",
                "type": "markdown",
              },
              "value": "*hello world*",
            }
          }
        />
      </div>,
      "className": "controls",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="label"
          >
            akey
          </div>,
          <ShortTextEditorComponent
            content={
              Object {
                "key": "akey",
                "short": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              }
            }
          />,
        ],
        "className": "control control-full",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "akey",
            "className": "label",
          },
          "ref": null,
          "rendered": "akey",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": Object {
              "key": "akey",
              "short": Object {
                "title": "thetitle",
                "type": "markdown",
              },
              "value": "*hello world*",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="control control-full"
        >
          <div
            className="label"
          >
            akey
          </div>
          <ShortTextEditorComponent
            content={
              Object {
                "key": "akey",
                "short": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              }
            }
          />
        </div>,
        "className": "controls",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="label"
            >
              akey
            </div>,
            <ShortTextEditorComponent
              content={
                Object {
                  "key": "akey",
                  "short": Object {
                    "title": "thetitle",
                    "type": "markdown",
                  },
                  "value": "*hello world*",
                }
              }
            />,
          ],
          "className": "control control-full",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "akey",
              "className": "label",
            },
            "ref": null,
            "rendered": "akey",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "content": Object {
                "key": "akey",
                "short": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> short:markdown -> ShortMarkdownEditor 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShortTextEditorComponent
    content={
      Object {
        "key": "akey",
        "short": Object {
          "title": "thetitle",
          "type": "markdown",
        },
        "value": "*hello world*",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "onChange": [Function],
      "type": "text",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "onChange": [Function],
        "type": "text",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:html -> LongHTMLEditor 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "html",
        },
        "value": "*hello world*",
      }
    }
    hideLabel={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="control control-full"
      >
        <div
          className="label"
        >
          akey
        </div>
        <EditorComponent
          content={
            Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "html",
              },
              "value": "*hello world*",
            }
          }
        />
      </div>,
      "className": "controls",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="label"
          >
            akey
          </div>,
          <EditorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "html",
                },
                "value": "*hello world*",
              }
            }
          />,
        ],
        "className": "control control-full",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "akey",
            "className": "label",
          },
          "ref": null,
          "rendered": "akey",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "html",
              },
              "value": "*hello world*",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="control control-full"
        >
          <div
            className="label"
          >
            akey
          </div>
          <EditorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "html",
                },
                "value": "*hello world*",
              }
            }
          />
        </div>,
        "className": "controls",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="label"
            >
              akey
            </div>,
            <EditorComponent
              content={
                Object {
                  "key": "akey",
                  "text": Object {
                    "title": "thetitle",
                    "type": "html",
                  },
                  "value": "*hello world*",
                }
              }
            />,
          ],
          "className": "control control-full",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "akey",
              "className": "label",
            },
            "ref": null,
            "rendered": "akey",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "content": Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "html",
                },
                "value": "*hello world*",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:html -> LongHTMLEditor 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditorComponent
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "html",
        },
        "value": "*hello world*",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <QuillComponent
        content={
          Object {
            "key": "akey",
            "text": Object {
              "title": "thetitle",
              "type": "html",
            },
            "value": "*hello world*",
          }
        }
      />,
      "className": "editor",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "content": Object {
          "key": "akey",
          "text": Object {
            "title": "thetitle",
            "type": "html",
          },
          "value": "*hello world*",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <QuillComponent
          content={
            Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "html",
              },
              "value": "*hello world*",
            }
          }
        />,
        "className": "editor",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "content": Object {
            "key": "akey",
            "text": Object {
              "title": "thetitle",
              "type": "html",
            },
            "value": "*hello world*",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:markdown -> LongMarkdownEditor 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "markdown",
        },
        "value": "*hello world*",
      }
    }
    hideLabel={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="control control-full"
      >
        <div
          className="label"
        >
          akey
        </div>
        <CodeMirrorComponent
          content={
            Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "markdown",
              },
              "value": "*hello world*",
            }
          }
        />
      </div>,
      "className": "controls",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="label"
          >
            akey
          </div>,
          <CodeMirrorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              }
            }
          />,
        ],
        "className": "control control-full",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "akey",
            "className": "label",
          },
          "ref": null,
          "rendered": "akey",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "markdown",
              },
              "value": "*hello world*",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="control control-full"
        >
          <div
            className="label"
          >
            akey
          </div>
          <CodeMirrorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              }
            }
          />
        </div>,
        "className": "controls",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="label"
            >
              akey
            </div>,
            <CodeMirrorComponent
              content={
                Object {
                  "key": "akey",
                  "text": Object {
                    "title": "thetitle",
                    "type": "markdown",
                  },
                  "value": "*hello world*",
                }
              }
            />,
          ],
          "className": "control control-full",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "akey",
              "className": "label",
            },
            "ref": null,
            "rendered": "akey",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "content": Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "markdown",
                },
                "value": "*hello world*",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:markdown -> LongMarkdownEditor 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CodeMirrorComponent
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "markdown",
        },
        "value": "*hello world*",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "codemirror",
    "nodeType": "host",
    "props": Object {
      "children": <textarea />,
      "className": "codemirror codemirror-long",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "textarea",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "codemirror",
      "nodeType": "host",
      "props": Object {
        "children": <textarea />,
        "className": "codemirror codemirror-long",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "textarea",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:text -> LongTextEditor 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "text",
        },
        "value": "*hello world*",
      }
    }
    hideLabel={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="control control-full"
      >
        <div
          className="label"
        >
          akey
        </div>
        <TextEditorComponent
          content={
            Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "text",
              },
              "value": "*hello world*",
            }
          }
        />
      </div>,
      "className": "controls",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="label"
          >
            akey
          </div>,
          <TextEditorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "text",
                },
                "value": "*hello world*",
              }
            }
          />,
        ],
        "className": "control control-full",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "akey",
            "className": "label",
          },
          "ref": null,
          "rendered": "akey",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": Object {
              "key": "akey",
              "text": Object {
                "title": "thetitle",
                "type": "text",
              },
              "value": "*hello world*",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="control control-full"
        >
          <div
            className="label"
          >
            akey
          </div>
          <TextEditorComponent
            content={
              Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "text",
                },
                "value": "*hello world*",
              }
            }
          />
        </div>,
        "className": "controls",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="label"
            >
              akey
            </div>,
            <TextEditorComponent
              content={
                Object {
                  "key": "akey",
                  "text": Object {
                    "title": "thetitle",
                    "type": "text",
                  },
                  "value": "*hello world*",
                }
              }
            />,
          ],
          "className": "control control-full",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "akey",
              "className": "label",
            },
            "ref": null,
            "rendered": "akey",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "content": Object {
                "key": "akey",
                "text": Object {
                  "title": "thetitle",
                  "type": "text",
                },
                "value": "*hello world*",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`<Editor> text:text -> LongTextEditor 2`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextEditorComponent
    content={
      Object {
        "key": "akey",
        "text": Object {
          "title": "thetitle",
          "type": "text",
        },
        "value": "*hello world*",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <textarea
        onChange={[Function]}
      />,
      "className": ".text",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "onChange": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": null,
      "type": "textarea",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <textarea
          onChange={[Function]}
        />,
        "className": ".text",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "textarea",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
